apps/dashboard/src/lib/validators/uptime-module.ts

```diff
@@ -0,0 +1,112 @@
+import { BehaviorSubject } from 'rxjs'
+import {
+  getValidatorUptimeSinceDate as getValidatorUptimeSinceDateOriginal,
+  uptimePeriodDefinition,
+  type UptimeValue
+} from '@api/utils/entities/component/validator'
+import BigNumber from 'bignumber.js'
+import type { ValidatorCollectionItem } from '@common/gateway-sdk'
+import { YEARLY_XRD_EMISSIONS } from '@constants'
+
+/**
+ * Uptime Data is an object which contains validator->uptime mapping for a specific uptime period.
+ *
+ * Example of Uptime Data:
+ * {
+ *  '1day': {
+ *     'validator_xyz': 0.99,
+ *     'validator_xxx': 0.54,
+ *   },
+ *  '1month': {
+ *     'validator_abc': 0.99,
+ *     'validator_bcd': 0.54,
+ *   },
+ * }
+ */
+export type UptimeData = Partial<
+  Record<UptimeValue, Record<string, number | undefined>>
+>
+
+export type UptimeModule = ReturnType<typeof UptimeModule>
+export const UptimeModule = (
+  dependencies: {
+    getValidatorUptimeSinceDate: typeof getValidatorUptimeSinceDateOriginal
+  } = {
+    getValidatorUptimeSinceDate: getValidatorUptimeSinceDateOriginal
+  }
+) => {
+  const { getValidatorUptimeSinceDate } = dependencies
+  const isLoading = new BehaviorSubject<boolean>(false)
+  let validators:
+    | Promise<{ address: string; totalStakeInXRD: BigNumber }[]>
+    | undefined
+  let uptimeData: UptimeData = {}
+  let totalAmountStaked: BigNumber | undefined
+  return {
+    setValidators: (
+      promise: Promise<{ address: string; totalStakeInXRD: BigNumber }[]>
+    ) => {
+      validators = promise
+      promise.then((v) => {
+        totalAmountStaked = v.reduce(
+          (prev, cur) => prev.plus(cur.totalStakeInXRD),
+          new BigNumber(0)
+        )
+      })
+    },
+    getDataForUptime: (uptime: UptimeValue) => uptimeData[uptime] || {},
+    getApy: (validator: ValidatorCollectionItem, uptime: UptimeValue) => {
+      if (!totalAmountStaked || totalAmountStaked.isZero()) {
+        throw new Error('Invalid totalAmountStaked')
```

Comment:
what's the reason why we're not using neverthrow?

Reply:
Hm I think I didn't want this additional complexity when using `getApy` function. This error only happens if there's actual bug in the code somewhere so I thought throwing is ok

Reply:
I can return `undefined` instead of throwing 🤷‍♂️